crypto/rand/rand_lib.c

```diff
@@ -90,15 +91,66 @@ int rand_rdcpu(void)
 }
 #endif
 
+/*
+ */
+static size_t entropy_from_system(RAND_DRBG *drbg,
+                                  unsigned char **pout,
+                                  int entropy, size_t min_len, size_t max_len)
+{
+    entropy /= 8;
+
+    if (rand_drbg.filled) {
+        /* Re-use what we have. */
+        *pout = drbg->randomness;
+        return sizeof(drbg->randomness);
+    }
+
+    /* If we don't have enough, get more. */
+    CRYPTO_THREAD_write_lock(rand_bytes.lock);
+    if (rand_bytes.curr < entropy) {
```

Comment:
A thought:

```
for (i=0; rand_bytes.curr < entropy && i < 10; i++)
```

Adding robustness when another process steals the bits added by `RAND_poll` before we grab the lock anew.

For 10 in an arbitrary integer base.

Anyway, just a thought.  The caller is going to have to deal with a zero return regardless.